product_profile/product.py

```diff
@@ -80,56 +80,48 @@ def _get_profile_fields(self):
-                if profile_obj._fields[field].type == 'many2many':
-                    profile[field] = [(6, 0, value)]
-            return profile
-        else:
-            return {field: None for field in fields}
+        vals = profile_obj.browse(profile_id).read(fields)[0]
+        vals.pop('id')
+        for field, value in vals.items():
+            if value and profile_obj._fields[field].type == 'many2one':
+                # m2o value is a tuple
+                vals[field] = value[0]
+            if profile_obj._fields[field].type == 'many2many':
+                vals[field] = [(6, 0, value)]
+            if PROF_DEFAULT_STR == field[:LEN_DEF_STR]:
+                vals[field[LEN_DEF_STR:]] = vals[field]
+                # prefixed fields must be removed from dict
+                # because they are in profile not in product
+                vals.pop(field)
+        return vals
 
     @api.onchange('profile_id')
     def _onchange_from_profile(self):
         """ Update product fields with product.profile corresponding fields """
-        for field, value in self._get_profile_data(self.profile_id.id).items():
-            try:
-                if PROF_DEFAULT_STR == field[:LEN_DEF_STR]:
-                    self[field[LEN_DEF_STR:]] = (
-                        self.profile_id[field])
-                else:
-                    self[field] = self.profile_id[field]
-            except ValueError as e:
-                raise UserError(format_except_message(e, field, self))
-            except Exception as e:
-                raise UserError("%s" % e)
+        if self.profile_id:
+            values = self._get_profile_data(self.profile_id.id)
+            for field, value in values.items():
+                try:
+                    self[field] = value
+                except Exception as e:
+                    raise UserError(format_except_message(e, field, self))
 
     @api.model
     def create(self, vals):
         if vals.get('profile_id'):
-            vals.update(
-                self._get_profile_data(vals['profile_id'], vals.keys()))
-        return super(ProductMixinProfile, self).create(
-            {k: v for k, v in vals.items() if PROF_DEFAULT_STR not in k})
+            vals.update(self._get_profile_data(vals['profile_id']))
+        return super(ProductMixinProfile, self).create(vals)
 
     @api.multi
     def write(self, vals):
         if vals.get('profile_id'):
-            vals.update(
-                self._get_profile_data(vals['profile_id'], vals.keys()))
-        return super(ProductMixinProfile, self).write(
-            {k: v for k, v in vals.items() if PROF_DEFAULT_STR not in k})
-
-    @api.model
-    def _get_profiles_to_filter(self):
-        """ Inherit if you want that some profiles doesn't have a filter """
-        return [(x.id, x.name) for x in self.env['product.profile'].search([])]
+            vals.update(self._get_profile_data(vals['profile_id']))
```

Comment:
As you said:
> With this, you're making that the product is updated each time you save it to the latest profile specs. I think this can provoke a lot of side effects

You're right, but if I put a button and somebody forget to click the button, then you have side effects: with the same profile applied you get records with different settings on fields which must be control by profile.

Instead of button, I propose to have a red text on product.profile form which explain the impact of the update of the profile. What do you think ?